### PR TITLE
Deprecate #[julia_pyo3] in favour of #[julia] + PyO3's own attributes (#275 Phase 3)

### LINKED ISSUE #275: Bind existing PyO3 crates from Julia without a RustCall-specific attribute
## Motivation

Today a crate must be annotated with RustCall's own `#[julia]` / `#[julia_pyo3]` to be usable from Julia. That excludes every existing crate that already exposes a Rust API through PyO3, and `#[julia_pyo3]` duplicates PyO3's job (with an ABI that currently differs between inline and crate mode, #269).

Observation: **PyO3's attribute macros keep the original Rust item.** In pyo3 0.29, `#[pyfunction]`, `#[pymethods]` and `#[pyclass]` all expand to `quote!(#ast #expanded)` (`pyo3-macros/src/lib.rs:154, 234, 199`), so the plain Rust function, impl block and struct survive with their original names, signatures and visibility. Nothing about them is Python-specific from the caller's side.

So RustCall does not need a new attribute at all: it can **discover** items through PyO3's attributes and generate the `extern "C"` shims in the wrapper crate it already builds for `@rust_crate` (today `generate_wrapper_lib_rs` only re-exports, assuming the target crate already carries `#[no_mangle]` symbols).

### Item-level proc-macro attributes

| Attribute | Applies to | Relevance for Julia bindings |
| --- | --- | --- |
| `#[pyfunction]` | `fn` | **Primary discovery marker.** The original `fn` is untouched; wrap it when the signature is FFI-representable. Options: `name`, `signature`, `text_signature`, `pass_module`, `crate`, `warn`. |
| `#[pymethods]` | `impl` block | **Primary marker for methods.** The original impl is untouched; wrap the inherent methods. |
| `#[pyclass]` | `struct` / `enum` | Marks the type that owns those methods. The struct is untouched but is *not* `#[repr(C)]`, so it must be handled as an opaque boxed pointer with accessors (what crate mode already does). Options: `name`, `module`, `crate`, `get_all`, `set_all`, `frozen`, `subclass`, `extends`, `eq`, `eq_int`, `ord`, `hash`, `str`, `sequence`, `mapping`, `dict`, `weakref`, `unsendable`, `immutable_type`, `freelist`, `rename_all`, `generic`, `from_py_object`, `skip_from_py_object`. |
| `#[pymodule]` | `fn` / `mod` | Python module assembly only. Useful as a hint for which items are public API; nothing to wrap. Options: `name`, `module`, `submodule`, `crate`, `gil_used`. |

### Method-level attributes (inside `#[pymethods]`)

`#[new]`, `#[staticmethod]`, `#[classmethod]`, `#[getter]`, `#[setter]`, `#[deleter]`, `#[classattr]`, plus `#[pyo3(name = ..., signature = ..., text_signature = ..., warn(...))]`. Mapping: `#[new]` → Julia constructor, `#[staticmethod]` → module-level function, `#[getter]`/`#[setter]` → property accessors, `#[classmethod]`/`#[classattr]`/`#[deleter]` → out of scope (they need a Python class object). Dunder methods (`__len__`, `__str__`, …) are ordinary Rust methods and can be wrapped or skipped by name.

### Field-level attributes (inside a `#[pyclass]` struct)

`#[pyo3(get)]`, `#[pyo3(set)]`, `#[pyo3(name = "...")]` — a direct match for RustCall's field getter/setter generation.

### Derives

`FromPyObject`, `IntoPyObject`, `IntoPyObjectRef` with their `#[pyo3(...)]` container/field options (`transparent`, `item`, `attribute`, `from_item_all`, `rename_all`, `into_py_with`, `default`, `crate`). These describe Python conversions only; irrelevant for Julia, but their presence is a useful signal that a type is a data-transfer type.

## Proposed design

1. **Discovery, not transformation.** Extend `rustcall_core` with a `pyo3` crate-scanning mode: `Attribute::Pyfunction` / `Pymethods` / `Pyclass` in the manifest model, recorded alongside the Rust name and the Python name from `#[pyo3(name = ...)]`.
2. **Wrapper crate generates calls, not re-exports.** `create_wrapper_crate` emits `#[no_mangle] pub extern "C" fn <crate>_<fn>(...)` shims that call `target_crate::path::fn(...)`, reusing the existing `rustcall_core::codegen` lowering (including the `(ptr, len)` string ABI and the `CResult`/`COption` wrappers). This is the piece that is missing today.
3. **Signature filter, fail closed.** Only wrap signatures that are FFI-representable. Skip and report, with a reason in the manifest, anything mentioning `Python<'py>`, `Bound<'py, T>`, `Py<T>`, `PyObject`, `PyRef`/`PyRefMut`, `&PyAny`, `PyResult<T>`, `#[pyo3(pass_module)]`, `#[pyo3(signature = ...)]` with `*args`/`**kwargs`, or `async fn`.
4. **Julia surface.** No new syntax: `@rust_crate "path/to/crate"` picks the items up. A keyword (`pyo3 = true/false`, default auto-detect from the crate's dependencies) decides whether PyO3 attributes are treated as discovery markers.

## Known constraints (to validate with an MWE first)

- **Visibility.** Only `pub` items are reachable from the wrapper crate. Many extension crates write `#[pyfunction] fn foo(...)` without `pub`; those cannot be wrapped from outside and must be reported clearly (a future in-tree mode could generate the shims inside the crate instead).
- **`PyResult<T>`.** `PyErr`'s `Display` needs the GIL, so an error cannot be turned into a Julia-side message without initializing Python. Out of scope for the first iteration.
- **Linking.** If the crate enables pyo3's `extension-module`, the cdylib deliberately leaves the Python symbols undefined, and `dlopen` with `RTLD_NOW` fails on Linux. The wrapper build must therefore disable that feature (or the crate's optional `python` feature) so no libpython is involved; when it cannot, this must be a clear error rather than a crash at call time.
- **`#[pyclass]` layout.** Not `#[repr(C)]`: always box and expose accessors, never map fields directly.

## Plan

- **Phase 0 (MWE).** A minimal crate with `#[pyfunction]` / `#[pyclass]` / `#[pymethods]`, a hand written wrapper crate that calls it, and a Julia script that loads the result — to confirm the linking story and the visibility constraint before building anything.
- **Phase 1.** Manifest + scanning: recognize the PyO3 attributes, record the Rust and Python names, classify each item as wrappable or skipped-with-reason.
- **Phase 2.** Wrapper-crate codegen that calls the target crate, reusing the existing lowering; feature handling for the pyo3 dependency.
- **Phase 3.** Deprecate `#[julia_pyo3]` in favour of plain `#[pyfunction]` + `#[julia]` (closes the ABI divergence in #269).

Related: #269 (`#[julia_pyo3]` ABI differs between inline and crate mode), #242 (String FFI, prerequisite for most real PyO3 signatures).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by terasakisatoshi**
## Summary

Phase 3 of #275: **deprecate `#[julia_pyo3]`** in favour of `#[julia]` stacked with PyO3's own attributes. Nothing that uses the attribute stops working; everything that uses it is told, at the right places, what to write instead.

`#[julia]` is additive since #279 — it keeps the annotated item exactly as written and emits the `extern "C"` entry point next to it — and PyO3's attributes keep the item too, so the two compose on one definition. That removes the reason `#[julia_pyo3]` existed, and with it the ABI divergence #269 described: the attribute exported `String` / `Result` signatures *as written* while `#[julia]` lowers them to the shared `(ptr, len)` / `CResult_*` ABI. Rather than extend the attribute into a second copy of `#[julia]`, it is frozen and deprecated.

## What changes

- **`deps/juliacall_macros`** — the `julia_pyo3` proc-macro carries `#[deprecated(note = …)]`, so rustc reports ``use of deprecated macro `julia_pyo3` `` at every use site with the replacement in the note. The expansion is unchanged.
- **Julia side** — `@rust_crate` / `write_bindings_to_file` warn once per crate whose scan contains `julia_pyo3` items (`_warn_deprecated_attributes`); `RustCall.scan_report` marks each such item. Binding is unchanged.
- **`examples/sample_crate_pyo3`** — migrated to the recommended shape: `#[julia] #[cfg_attr(feature = "python", pyo3::pyfunction)]` on functions, `#[julia] #[cfg_attr(feature = "python", pyo3::pyclass(get_all, set_all))]` on the struct, `#[julia]` on the impl and its methods, and a `#[cfg(feature = "python")] #[pyo3::pymethods] impl` written as PyO3 code that delegates to the Rust methods (pyo3's inner attributes cannot be gated with `cfg_attr`, and a type has one inherent method per name). Builds and passes `cargo test` with the feature off, `cargo check` / `clippy -D warnings` with `--features python`; the Julia demo test still runs it. README / `main.jl` / `main.py` updated.
- **Docs** — `docs/src/pyo3.md` gains "Migrating from `#[julia_pyo3]`" (mapping table, the two facts about the impl-block row, what changes on the ABI side); the stale "unchanged by #275" note is gone. `rustcall_core` doc comments say the flavour is frozen and deprecated.
- **CHANGELOG** — a `### Deprecated` entry.
- **Removal** is a breaking change and is tracked separately: #312 (to batch with #265 Phase 2 / #300).

## Tests

| deliverable | test |
| --- | --- |
| rustc warns at every `#[julia_pyo3]` use site (function, struct, impl) with the replacement and the issue in the note | `deps/juliacall_macros/tests/deprecated.rs` — nested `cargo check --message-format=json` on `tests/fixtures/uses_julia_pyo3`, asserting exactly three `use of deprecated macro` warnings |
| the attribute still expands and its manifest is unchanged | existing `deps/rustcall_core/tests/corpus/pyo3.rs` goldens (unchanged) and the `#[julia_pyo3]` string-ABI testset in `test/test_julia_attribute.jl` (unchanged) |
| `@rust_crate` / `write_bindings_to_file` warn once per crate, and are silent without the attribute | `test/test_julia_attribute.jl`, "`#[julia_pyo3]` is deprecated" block (`@test_logs`) |
| the migrated example binds and behaves as before from Julia | `test/test_crate_bindings.jl`, "sample_crate_pyo3 Julia Demo" |
| `#[julia]` stacks with `#[pyfunction]` | `deps/juliacall_macros/examples/pyo3_compose.rs` (compiled by `cargo test`, #279) and `tests/additive.rs` (unchanged) |

## Issue status

Closes #275 — Phase 0 (MWE, recorded on the issue), Phase 1 (#294), Phase 2 (#307), Phase 3 (this PR). The scan/link-plan edge cases those phases surfaced live in #303; module-qualified symbols in #300; the attribute's removal in #312.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_014iJ1dZ7KEebWDjdY7fhPbw

### CODE REVIEW

**Comment by terasakisatoshi** [comment]
@codex review

File: src/crate_bindings.jl [hunk 1]
```diff
@@ -237,6 +237,27 @@ function crate_lib_root(crate_path::AbstractString, cargo_toml::AbstractDict)
...
+
+Warn, once per crate, when the scan found items produced by a deprecated
+RustCall attribute — today `#[julia_pyo3]` (#275 Phase 3). The items are still
+bound exactly as before; the warning is the Julia-side counterpart of the
+`use of deprecated macro` rustc reports at each use site, for users who only
+ever see the build through `@rust_crate`. Returns the number of such items.
+"""
+function _warn_deprecated_attributes(info::CrateInfo)
+    deprecated = count(f -> f.attribute === :julia_pyo3, info.julia_functions) +
+                 count(s -> s.attribute === :julia_pyo3, info.julia_structs)
```
[the remark this answers is not in the document]
Comment (terasakisatoshi) [reply src/crate_bindings.jl]: Fixed in 64d1437: a method's manifest entry now records the attribute of the impl block it came from (`Method.attribute` — `julia`, `julia_pyo3`, `py_methods`; omitted for an inline-mode impl; additive within schema 6 like `Method.cfg`), `RustMethod` reads it, and `_uses_julia_pyo3` decides per struct from the struct's own attribute *or* any method's — so `_warn_deprecated_attributes` and the `scan_report` marker both catch a `#[julia] struct` whose impl is the deprecated one. Test: the `Gauge` / `Clean` cases in `test/test_julia_attribute.jl` (the method carries `:julia_pyo3`, the struct counts, a `#[julia] impl` does not).

**Comment by terasakisatoshi** [comment]
@codex review

File: src/crate_bindings.jl [hunk 2]
```diff
@@ -1574,6 +1611,7 @@ function generate_bindings(crate_path::String;
     @info "Scanning crate at $crate_path"
     info = scan_crate(crate_path)
     @info "Found $(length(info.julia_functions)) functions and $(length(info.julia_structs)) structs"
+    _warn_deprecated_attributes(info)
```
[the remark this answers is not in the document]
Comment (terasakisatoshi) [reply src/crate_bindings.jl]: Fixed in 1528f7e: both entry points now call `_warn_deprecated_attributes` after `_plain_scan_info`, on the resolved scan the bindings are emitted from, so a `#[cfg_attr(feature = "legacy", julia_pyo3)]` is seen exactly under the build that enables it. Test: the feature-gated case in `test/test_julia_attribute.jl` — resolved with `legacy` off the item is `:julia` and nothing is said; with `legacy` on it is `:julia_pyo3` and the warning fires.

**Comment by terasakisatoshi** [comment]
@codex review

File: src/crate_bindings.jl [hunk 3]
```diff
@@ -1607,6 +1643,10 @@ function generate_bindings(crate_path::String;
         end
     end
     info = _plain_scan_info(crate_path, info, features, default_features, build_release)
+    # On the *resolved* scan: a `#[cfg_attr(feature = "legacy", julia_pyo3)]`
+    # is an attribute only under the build that enables it, which the lenient
+    # scan leaves undecided (#314 review).
+    _warn_deprecated_attributes(info)
```
[the remark this answers is not in the document]
Comment (terasakisatoshi) [reply src/crate_bindings.jl:1654]: Fixed in 3aba799: both entry points now decide the notice from the crate's own scan under the wrapper's build (`_scan_under_plan(crate_path, info, wrapper.plan)`) *before* handing over to the wrapper — `generate_bindings` before its early return, `write_bindings_to_file` before `info` becomes the wrapper's manifest. Regression: `examples/sample_crate_pyo3_mixed` keeps one `#[julia_pyo3]` item on purpose and the mixed testset in `test/test_pyo3_wrapper.jl` asserts the warning from `generate_bindings` and from `write_bindings_to_file`, plus that the item is bound.

**Comment by terasakisatoshi** [comment]
@codex review

File: src/crate_bindings.jl [hunk 4]
```diff
@@ -237,6 +237,42 @@ function crate_lib_root(crate_path::AbstractString, cargo_toml::AbstractDict)
...
+bound exactly as before; the warning is the Julia-side counterpart of the
+`use of deprecated macro` rustc reports at each use site, for users who only
+ever see the build through `@rust_crate`. Returns the number of such items.
+"""
+function _warn_deprecated_attributes(info::CrateInfo)
+    # A struct counts once whether the attribute sits on the struct itself or
+    # only on one of its impl blocks (`RustMethod.attribute`): a `#[julia]
+    # struct` with a `#[julia_pyo3] impl` is a use of the deprecated macro too.
+    deprecated = count(f -> f.attribute === :julia_pyo3, info.julia_functions) +
+                 count(s -> _uses_julia_pyo3(s), info.julia_structs)
```
[the remark this answers is not in the document]
Comment (terasakisatoshi) [reply src/crate_bindings.jl:254]: Tracked as #315 rather than fixed here. The root cause is a pre-existing limitation of the `#[julia]` crate scan, not of this PR: an `impl` in another module than its struct is never attached (`collect_struct_models_in` works per module level), so the layout you describe loses the methods' *bindings* entirely — the missing deprecation notice is the smallest consequence, and rustc's `use of deprecated macro` still fires at that impl. The fix is to collect structs and impls crate-wide and marry them by resolved path, as the PyO3 scan already does for `#[pymethods]` (`Pyo3Scan::finish`, `PathQualifier`); #315 carries the split-layout regression fixture as an acceptance criterion, which will exercise the notice too.
